### ISSUE TITLE
More lenient jsx-no-multiline-js rule

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint-react#50]
I think the rule should be lenient about these kinds of multiline expressions because they don't hinder readability too much:

![image](https://cloud.githubusercontent.com/assets/723999/21246519/5050b98a-c2f7-11e6-90ad-c40f56b33b48.png)

update: we should also add more exemptions:

- `allow-function-call` to ignore function call expressions
- `allow-literal` to ignore checking object and array literals
- `allow-element` to ignore JSX elements

**Comment by markwongsk** [comment]
How lenient do you think it should be? Should we just allow object literals, or do you think this should allow things like
```
popoverProps={this.computePopoverProps({
    popoverClassName: "docs-dropdown-ramen-popover",
})}
```
Also, this rule currently doesn't care about spread attributes, eg
```
<Dropdown {...{
    items: mocks.groups(),
    placeholder: "Ramen joint",
}} />
```
but that's probably a separate issue.

**Comment by piotrwitek** [comment]
@adidahiya @markwongsk this is hanging for quite some time, can we push for this change?

I really need this configurability and I think that disable for object literals is enough for a start.
Checking tslint conventions suggested name might be: "ignore-object-literal"

**Comment by adidahiya** [comment]
@markwongsk I guess we could break down the use cases into "be lenient on multiline functions / lambdas" and "be lenient on object or array literals". so this rule could have two new options: `ignore-functions` and `ignore-literals`

@piotrwitek open to PRs, as always :)

**Comment by MiguelMadero** [comment]
Would it be a simpler distinction between JS statements and expressions? It seems like multline expressions (object literals, ternary, function calls, etc) make more sense than multiple statements.

**Comment by adidahiya** [comment]
@MiguelMadero eh, that wouldn't really solve my use case because so much of JS syntax is a valid expression. My primary motivation for using this rule was to prevent multiline ternarys and function calls (primarily `.map`, since it's used to generate lists)

```ts
// bad
<div className={
    someComplicatedCondition
        ? "positive-case"
        : "positive-case"}
    otherProp={true}
/>

// bad
<div>
    {this.props.foo.map((f) => {
        return <span>{f}</span>;
    }
</div>
```

I could imagine another option here for stylistic preference, however, `ignore-conditional-expressions`, to allow multiline ternarys like in my first example ^

**Comment by pingshunhuang** [comment]
@adidahiya Could you please recommend a way for the 2 cases you gave above. I got this rules triggered a lot and I'm not sure how to improve them. Thanks.

**Comment by petrdanecek** [comment]
Is there any plan or update about this topic?

**Comment by adidahiya** [comment]
Closing due to deprecation, see #210
